### PR TITLE
feat(emqx_auth): implement API to re-order all authenticators/authz sources

### PR DESCRIPTION
**Pull request by SergeTupchiy**
Fixes EMQX-11770

Release version: `v/e5.6.0`

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [x] Added tests for the changes
- [x] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [x] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [x] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [x] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

**Comment by SergeTupchiy** [comment]
API:

```
PUT /authentication​/order
[
  {"id": "password_based:built_in_database"},
  {"id": "password_based:postgresql"}
]
```

```
PUT /authorization/sources/order
[
  {"type": "file"},
  {"type": "built_in_database"}
]
```

File: apps/emqx_auth/src/emqx_authz/emqx_authz.erl [hunk 1]
```diff
@@ -258,6 +262,31 @@ do_pre_config_update({?CMD_REPLACE, Sources}, _OldSources) ->
...
+                case lists:keytake(Type, 1, RemOldSourcesAcc) of
+                    {value, {_Type, Source}, RemOldSourcesAcc1} ->
+                        {[Source | NewSourcesAcc], NotFoundTypesAcc, RemOldSourcesAcc1};
+                    false ->
+                        {NewSourcesAcc, [Type | NotFoundTypesAcc], RemOldSourcesAcc}
+                end
+            end,
+            {[], [], OldSourcesWithTypes},
+            NewSourcesOrder1
+        ),
```
Comment (zmstone) [inline apps/emqx_auth/src/emqx_authz/emqx_authz.erl:280]: nit: impl a loop function.

File: apps/emqx_auth/src/emqx_authn/emqx_authn_api.erl [hunk 2]
```diff
@@ -1253,6 +1292,19 @@ serialize_error({unknown_authn_type, Type}) ->
         code => <<"BAD_REQUEST">>,
         message => binfmt("Unknown type '~p'", [Type])
     }};
+serialize_error(#{not_found := NotFound, not_reordered := NotReordered}) ->
+    NotFoundFmt = "Authenticators: ~p are not found",
```
Comment (zmstone) [inline apps/emqx_auth/src/emqx_authn/emqx_authn_api.erl:1296]: will it produce `[<<"redis">>, <<"mysql">>]`?
if yes, maybe better to use `[~ts]` for a `lists:join(",", NotFound)`

File: apps/emqx_auth/test/emqx_authz/emqx_authz_api_sources_SUITE.erl [hunk 3]
```diff
@@ -582,6 +582,121 @@ t_source_move(_) ->
 
     ok.
 
+t_sources_reorder(_) ->
```
Comment (zmstone) [inline apps/emqx_auth/test/emqx_authz/emqx_authz_api_sources_SUITE.erl:585]: nit: for regression purpose, maybe add a disabled source (the enable/disable state should not affect ordering).

File: rel/i18n/emqx_authz_api_sources.hocon [hunk 4]
```diff
@@ -35,6 +35,11 @@ authorization_sources_type_status_get.desc:
 authorization_sources_type_status_get.label:
 """Get a authorization source"""
 
+authorization_sources_order_put.desc:
+"""Reorder all authorization sources."""
+authorization_sources_order_put.label:
+"""Reorder authorization sources"""
```
Comment (zmstone) [inline rel/i18n/emqx_authz_api_sources.hocon:41]:
```suggestion
"""Reorder Authorization Sources"""
```

File: rel/i18n/emqx_authn_api.hocon [hunk 5]
```diff
@@ -60,6 +60,11 @@ authentication_post.desc:
 authentication_post.label:
 """Create authenticator"""
 
+authentication_order_put.desc:
+"""Reorder all authenticators in global authentication chain."""
+authentication_order_put.label:
+"""Reorder authenticators"""
```
Comment (zmstone) [inline rel/i18n/emqx_authn_api.hocon:66]:
```suggestion
"""Reorder Authenticators"""
```

File: apps/emqx_auth/src/emqx_authn/emqx_authn_config.erl [hunk 6]
```diff
@@ -135,6 +135,30 @@ do_pre_config_update(_, {move_authenticator, _ChainName, AuthenticatorID, Positi
 do_pre_config_update(ConfPath, {merge_authenticators, NewConfig}, OldConfig) ->
     MergeConfig = merge_authenticators(OldConfig, NewConfig),
     do_pre_config_update(ConfPath, MergeConfig, OldConfig);
+do_pre_config_update(_ConfPath, {reorder_authenticators, NewOrder}, OldConfig) ->
+    OldConfigWithIds = [{authenticator_id(Auth), Auth} || Auth <- OldConfig],
+    {NewConfig, NotFoundIds, RemOldConfig} =
+        lists:foldr(
```
Comment (zmstone) [inline apps/emqx_auth/src/emqx_authn/emqx_authn_config.erl:141]: nit: implement a loop function

**Comment by zmstone** [comment]
> API:
> 
> ```
> PUT /authentication​/order
> [
>   {"id": "password_based:built_in_database"},
>   {"id": "password_based:postgresql"}
> ]
> ```
> 
> ```
> PUT /authorization/sources/order
> [
>   {"type": "file"},
>   {"type": "built_in_database"}
> ]
> ```

Is this based on what's returned in the GET APIs ?

**Comment by SergeTupchiy** [comment]
> > API:
> > ```
> > PUT /authentication​/order
> > [
> >   {"id": "password_based:built_in_database"},
> >   {"id": "password_based:postgresql"}
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > ```
> > PUT /authorization/sources/order
> > [
> >   {"type": "file"},
> >   {"type": "built_in_database"}
> > ]
> > ```
> 
> Is this based on what's returned in the GET APIs ?

@zmstone , 
Not exactly, though "type" (authz) and "id" (authn) are returned in GET APIs the same way:
```
GET /authorization​/sources
{
  "sources": [
    {
      "enable": true,
      "type": "built_in_database"
    }
]}
```

```
GET /authentication
[
  {
    "id": "password_based:built_in_database",
    "backend": "built_in_database",
    "enable": true,
    "mechanism": "password_based",
    "password_hash_algorithm": {
      "name": "sha256",
      "salt_position": "suffix"
    },
    "user_id_type": "username"
  }
]
```

It's not possible to send the exact object returned from GET to PUT `../order`, since the latter expects only one field (id for authn, type for authz).

Should it be changed?

File: apps/emqx_auth/src/emqx_authn/emqx_authn_api.erl (hunk 2 above)
Comment (SergeTupchiy) [reply apps/emqx_auth/src/emqx_authn/emqx_authn_api.erl:1296]: You are right, thanks!

File: rel/i18n/emqx_authn_api.hocon (hunk 5 above)
Comment (SergeTupchiy) [reply rel/i18n/emqx_authn_api.hocon]: Fixed

File: apps/emqx_auth/test/emqx_authz/emqx_authz_api_sources_SUITE.erl (hunk 3 above)
Comment (SergeTupchiy) [reply apps/emqx_auth/test/emqx_authz/emqx_authz_api_sources_SUITE.erl:585]: Done for both authz and authn.

File: apps/emqx_auth/src/emqx_authn/emqx_authn_api.erl (hunk 2 above)
Comment (SergeTupchiy) [reply apps/emqx_auth/src/emqx_authn/emqx_authn_api.erl]: Done

File: apps/emqx_auth/src/emqx_authz/emqx_authz.erl (hunk 1 above)
Comment (SergeTupchiy) [reply apps/emqx_auth/src/emqx_authz/emqx_authz.erl]: Done

**Review by zmstone — APPROVED** [review]

**Pull request merged**